codegen/cmd/injection-gen/generators/reconciler_reconciler.go

```diff
@@ -0,0 +1,352 @@
+			Package: "k8s.io/client-go/util/retry",
+			Name:    "RetryOnConflict",
+		}),
+		"apierrsIsNotFound": c.Universe.Function(types.Name{
+			Package: "k8s.io/apimachinery/pkg/api/errors",
+			Name:    "IsNotFound",
+		}),
+		"metav1GetOptions": c.Universe.Function(types.Name{
+			Package: "k8s.io/apimachinery/pkg/apis/meta/v1",
+			Name:    "GetOptions",
+		}),
+		"zapSugaredLogger": c.Universe.Type(types.Name{
+			Package: "go.uber.org/zap",
+			Name:    "SugaredLogger",
+		}),
+	}
+
+	sw.Do(reconcilerInterfaceFactory, m)
+	sw.Do(reconcilerNewReconciler, m)
+	sw.Do(reconcilerImplFactory, m)
+	sw.Do(reconcilerStatusFactory, m)
+	sw.Do(reconcilerFinalizerFactory, m)
+
+	return sw.Error()
+}
+
+var reconcilerInterfaceFactory = `
+// Interface defines the strongly typed interfaces to be implemented by a
+// controller reconciling {{.type|raw}}.
+type Interface interface {
+	// ReconcileKind implements custom logic to reconcile {{.type|raw}}. Any changes
+	// to the objects .Status or .Finalizers will be propagated to the stored
+	// object. It is recommended that implementors do not call any update calls
+	// for the Kind inside of ReconcileKind, it is the responsibility of the calling
+	// controller to propagate those properties.
+	ReconcileKind(ctx context.Context, o *{{.type|raw}}) {{.reconcilerEvent|raw}}
+}
+
+// reconcilerImpl implements controller.Reconciler for {{.type|raw}} resources.
+type reconcilerImpl struct {
+	// Client is used to write back status updates.
+	Client {{.clientsetInterface|raw}}
+
+	// Listers index properties about resources
+	Lister {{.resourceLister|raw}}
+
+	// Recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	Recorder {{.recordEventRecorder|raw}}
+
+	// FinalizerName is the name of the finalizer to use when finalizing the
+	// resource.
+	FinalizerName string
+
+	// reconciler is the implementation of the business logic of the resource.
+	reconciler Interface
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*reconcilerImpl)(nil)
+
+`
+
+var reconcilerNewReconciler = `
+func NewReconciler(ctx context.Context, logger *{{.zapSugaredLogger|raw}}, client {{.clientsetInterface|raw}}, lister {{.resourceLister|raw}}, recorder {{.recordEventRecorder|raw}}, r Interface) {{.controllerReconciler|raw}} {
+	return &reconcilerImpl{
+		Client: client,
+		Lister: lister,
+		Recorder: recorder,
+		FinalizerName: defaultFinalizerName,
+		reconciler:    r,
+	}
+}
+`
+
+var reconcilerImplFactory = `
+// Reconcile implements controller.Reconciler
+func (r *reconcilerImpl) Reconcile(ctx context.Context, key string) error {
+	logger := {{.loggingFromContext|raw}}(ctx)
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := {{.cacheSplitMetaNamespaceKey|raw}}(key)
+	if err != nil {
+		logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+    // TODO: this is needed for serving.
```

Comment:
Seems important?

Reply:
Serving has a bunch of other problems, there will be updates to this code as more integration happens.